File: relax-core-spring-boot-starter/src/main/java/com/infilos/spring/config/ShutdownConfigure.java
package com.infilos.spring.config;

import com.infilos.utils.Loggable;
import org.eclipse.collections.impl.factory.Lists;
import org.springframework.context.ApplicationListener;
import org.springframework.context.event.ContextClosedEvent;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Objects;

@Component
public class ShutdownConfigure implements ApplicationListener<ContextClosedEvent>, Loggable {

  private static final List<Runnable> hooks = Lists.mutable.empty();

  public synchronized void register(Runnable hook) {
    if (Objects.nonNull(hook)) {
      hooks.add(hook);
    }
  }

  @Override
  public void onApplicationEvent(ContextClosedEvent contextClosedEvent) {
    log().info("Start invoke spring shutdown hooks, totally {}...", hooks.size());
    int failed = 0;

    for (Runnable hook : hooks) {
      try {
        hook.run();
      } catch (Throwable e) {
        failed += 1;
        log().error("Invoke spring shutdown hook failed.", e);
      }
    }

    log().info("Finish invoke spring shutdown hooks, totally {}, failed {}.", hooks.size(), failed);
  }
}

File: relax-scala-spring-boot-starter/src/main/scala/com/infilos/spring/config/ShutdownConfigure.scala
package com.infilos.spring.config

import com.infilos.spring.utils._
import org.springframework.context.ApplicationListener
import org.springframework.context.event.ContextClosedEvent

import java.util.concurrent.TimeUnit
import javax.annotation.PostConstruct
import scala.collection.mutable.ArrayBuffer

class ShutdownConfigure extends ApplicationListener[ContextClosedEvent] with Loggable {

  @PostConstruct
  def construct(): Unit = {
    // setup io-thread-pool shutdown hook
    ShutdownConfigure.register(() => {
      try {
        Executors.IOThreadPool.shutdown()
        if (!Executors.IOThreadPool.awaitTermination(30, TimeUnit.SECONDS)) {
          log.warn("IO thread pool shutdown gracefully failed.")
        } else {
          log.warn("IO thread pool shutdown gracefully succed.")
        }
      } catch {
        case _: Throwable => Thread.currentThread().interrupt()
      }
    })

    // setup db-thread-pool shutdown hook
    ShutdownConfigure.register(() => {
      try {
        Executors.DBThreadPool.shutdown()
        if (!Executors.IOThreadPool.awaitTermination(30, TimeUnit.SECONDS)) {
          log.warn("DB thread pool shutdown gracefully failed.")
        } else {
          log.warn("DB thread pool shutdown gracefully succed.")
        }
      } catch {
        case _: Throwable => Thread.currentThread().interrupt()
      }
    })
  }

  override def onApplicationEvent(e: ContextClosedEvent): Unit = {
    ShutdownConfigure.shutdowns.foreach { task =>
      try {
        task.run()
      } catch {
        case _: Throwable => Thread.currentThread().interrupt()
      }
    }
  }
}

object ShutdownConfigure {
  private val shutdowns: ArrayBuffer[Runnable] = ArrayBuffer.empty

  def register(task: Runnable): Unit = synchronized {
    shutdowns.append(task)
  }
}
